modules/javafx.graphics/src/main/native-glass/gtk/glass_dnd.cpp

```diff
@@ -760,132 +734,346 @@ static gboolean dnd_source_set_raw(GtkWidget *widget, GtkSelectionData *sel_data
-{
-    if (drag_widget) {
-        GdkDragAction action = gdk_drag_context_get_selected_action(context);
-        dnd_set_performed_action(translate_gdk_action_to_glass(action));
+static void process_dnd_source_grab_broken(DragSourceContext *ctx, GdkEvent *event) {
+    GdkEventGrabBroken *gb_event = &event->grab_broken;
+
+    // grabbed the same window
+    if (gb_event->implicit || gb_event->grab_window == ctx->dnd_window) {
+        return;
     }
-    gdk_threads_add_idle((GSourceFunc) dnd_destroy_drag_widget_callback, NULL);
+
+    gdk_drag_abort(ctx->dnd_ctx, GDK_CURRENT_TIME);
 }
 
-static gboolean dnd_drag_failed_callback(GtkWidget *widget,
-                                     GdkDragContext *context,
-                                     GtkDragResult result,
-                                     gpointer user_data)
-{
-    dnd_set_performed_action(com_sun_glass_ui_gtk_GtkDnDClipboard_ACTION_NONE);
-    gdk_threads_add_idle((GSourceFunc) dnd_destroy_drag_widget_callback, NULL);
+static bool process_dnd_source_mouse_release(DragSourceContext *ctx, GdkEvent *event) {
+    // already processed
+    if (ctx->device_released) {
+        in_drag = false;
+        return true;
+    }
 
-    return FALSE;
+    GdkDragAction selected = gdk_drag_context_get_selected_action(ctx->dnd_ctx);
+
+    if (selected) {
+        ctx->performed_action = translate_gdk_action_to_glass(selected);
+        gdk_drag_drop(ctx->dnd_ctx, GDK_CURRENT_TIME);
+    } else {
+        gdk_drag_abort(ctx->dnd_ctx, GDK_CURRENT_TIME);
+    }
+
+    ctx->device_released = true;
+
+    // if on the same window we must wait for drop do complete
+    if (ctx->dest_window != ctx->dnd_window) {
+        in_drag = false;
+    }
+
+    // the GDK_BUTTON_RELEASE will be re-inserted on the queue because
+    // WindowContext mouse release would stop DND before it's completed
+    gdk_display_put_event(gdk_display_get_default(), event);
+
+    return false;
 }
 
-static void dnd_data_get_callback(GtkWidget *widget,
-                                  GdkDragContext *context,
-                                  GtkSelectionData *data,
-                                  guint info,
-                                  guint time,
-                                  gpointer user_data)
-{
-    GdkAtom atom = gtk_selection_data_get_target(data);
-
-    if (target_is_text(atom)) {
-        dnd_source_set_string(widget, data, atom);
-    } else if (target_is_image(atom)) {
-        dnd_source_set_image(widget, data, atom);
-    } else if (target_is_uri(atom)) {
-        dnd_source_set_uri(widget, data, atom);
+static void process_drag_motion(DragSourceContext *ctx, gint x_root, gint y_root, guint state) {
+    GdkWindow *dest_window;
+    GdkDragProtocol prot;
+    GdkWindow *ignore = NULL;
+
+    ctx->last_x = x_root;
+    ctx->last_y = y_root;
+
+    if (ctx->drag_view) {
+        ctx->drag_view->move(x_root, y_root);
+        ignore = ctx->drag_view->get_window();
```

Comment:
This tells GDK to ignore the drag view as a possible destination because it's under the cursor.